### PR TITLE
Updated config.go - added defaults config option

### PR DESCRIPTION
**Pull request by JNickson**
Allows for subscription_id and resource_group to have a default value to fall back on per config file for all resources in that file.

### CODE REVIEW

**Comment by alec-pinson** [comment]
Haven't looked properly yet because it's hard to read on my phone but might be worth adding an app3.yaml in resources with an example of how to use what you added.

File: config.go [hunk 1]
```diff
@@ -10,18 +10,29 @@ import (
...
 	TTL         int                     `yaml":ttl"`
 }
 
+type Defaults struct {
+	SubscriptionId string `yaml:"subscription_id"`
+	ResourceGroup  string `yaml:"resource_group"`
+}
+
 type ResourceConfiguration struct {
+	Defaults       []string `yaml:"defaults"`
```
Comment (alec-pinson) [inline config.go]: Think you added this and don't actually use it ?

File: config.go [hunk 2]
```diff
@@ -10,18 +10,29 @@ import (
...
+}
+
 type Configuration struct {
 	File        string
 	Debug       bool
 	Url         string                  `yaml:"url"`
 	Redis       RedisConfiguration      `yaml:"redis"`
 	Auth        Authentication          `yaml:"auth"`
 	Resources   []ResourceConfiguration `yaml:"resources"`
+	Defaults    []Defaults              `yaml:"defaults"`
```
Comment (alec-pinson) [inline config.go]: Don't think there's any need for this to be an array of Defaults?

Comment (alec-pinson) [reply config.go]:
```suggestion
	Defaults    Defaults              `yaml:"defaults"`
```

File: config.go [hunk 3]
```diff
@@ -91,71 +104,86 @@ func (c *Configuration) load(reload ...bool) *Configuration {
...
-				cd.Name = resource.Name
-				cd.IPWhiteList = resource.IPWhiteList
-				cd.Group = resource.Group
-				cd.new(cd)
+	for _, config := range scf.Config {
+		subId := ""
+		rg := ""
+		for _, resource := range config.Resources {
+			if resource.SubscriptionId == "" {
+				subId = config.Defaults[0].SubscriptionId
```
Comment (alec-pinson) [inline config.go]:
```suggestion
				subId = config.Defaults.SubscriptionId
```
Don't think this needs to be an array

File: config.go [hunk 4]
```diff
@@ -91,71 +104,86 @@ func (c *Configuration) load(reload ...bool) *Configuration {
...
+			} else if resource.SubscriptionId != "" {
+				subId = resource.SubscriptionId
+			}
+			if resource.ResourceGroup == "" {
+				rg = config.Defaults[0].ResourceGroup
```
Comment (alec-pinson) [inline config.go]:
```suggestion
				rg = config.Defaults.ResourceGroup
```
Don't think this needs to be an array

File: config/config.yaml [hunk 5]
```diff
@@ -19,6 +19,10 @@ redis:
   port: 6379
   token: my-sup3r-comp1ic4t3d-s3cr3t-t0k3n
 
+defaults:
+  - subscription_id: notreal-default-subscription
+    resource_group: notreal-default-rg
```
Comment (alec-pinson) [inline config/config.yaml]:
```suggestion
defaults:
  subscription_id: notreal-default-subscription
  resource_group: notreal-default-rg
```
Don't think this needs to be an array/list/slince whatever it is :D

File: config/resources/app2.yaml [hunk 6]
```diff
@@ -1,3 +1,7 @@
+defaults:
+  - subscription_id: notreal-default-subscription
+    resource_group: notreal-default-rg
```
Comment (alec-pinson) [inline config/resources/app2.yaml]:
```suggestion
  subscription_id: notreal-default-subscription
  resource_group: notreal-default-rg
```
Don't think this needs to be an array/list/slince whatever it is :D

File: config/resources/app3.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+defaults:
+  - subscription_id: notreal-default-subscription
+    resource_group: notreal-default-rg
```
Comment (alec-pinson) [inline config/resources/app3.yaml]:
```suggestion
  subscription_id: notreal-default-subscription
  resource_group: notreal-default-rg
```
Don't think this needs to be an array/list/slince whatever it is :D

File: config/resources/app3.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    type: keyvault
+    name: app2notrealkeyvault
+
+# This resource has subscription and resource group so will override the default 
+  - cloud: azure
+    type: frontdoor
+    subscription_id: notreal-not-real-not-notreal
+    resource_group: notreal-rg
+    policy_name: notrealpolicy
+
```
Comment (alec-pinson) [inline config/resources/app3.yaml]:
```suggestion

```
Fix line end

File: config.go (hunk 2 above)
Comment (JNickson) [reply config.go]: ah yeah okay that works the same lol

File: config.go (hunk 1 above)
Comment (JNickson) [reply config.go]: yeah that was a mistake was mean't to remove that

File: config.go [hunk 9]
```diff
@@ -91,71 +103,86 @@ func (c *Configuration) load(reload ...bool) *Configuration {
...
-				cd.IPWhiteList = resource.IPWhiteList
-				cd.Group = resource.Group
-				cd.new(cd)
+	for _, config := range scf.Config {
+		subId := ""
+		rg := ""
+		for _, resource := range config.Resources {
+			if resource.SubscriptionId == "" {
+				subId = config.Defaults.SubscriptionId
+			} else if resource.SubscriptionId != "" {
```
Comment (alec-pinson) [inline config.go]:
```suggestion
			} else {
```

File: config.go [hunk 10]
```diff
@@ -91,71 +103,86 @@ func (c *Configuration) load(reload ...bool) *Configuration {
...
+				subId = resource.SubscriptionId
+			}
+			if resource.ResourceGroup == "" {
+				rg = config.Defaults.ResourceGroup
+			} else if resource.ResourceGroup != "" {
```
Comment (alec-pinson) [inline config.go]:
```suggestion
			} else {
```

**Review by alec-pinson — CHANGES_REQUESTED** [review]
Just a lil cleanup on the if else :)

**Pull request merged**